### PR TITLE
Simplify duplicate checks for mir validator

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This removes unnecessary allocations & is less code.

### CODE REVIEW

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 1]
```diff
@@ -994,16 +991,14 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                 // passed by a reference to the callee. Consequently they must be non-overlapping.
                 // Currently this simply checks for duplicate places.
                 self.place_cache.clear();
-                self.place_cache.push(destination.as_ref());
+                self.place_cache.insert(destination.as_ref());
+                let mut has_duplicates = false;
                 for arg in args {
                     if let Operand::Move(place) = arg {
-                        self.place_cache.push(place.as_ref());
+                        has_duplicates |= !self.place_cache.insert(place.as_ref());
```
Comment (klensy) [inline compiler/rustc_const_eval/src/transform/validate.rs:998]: Here can just break from `for` when dupe found, but given that this is error only code path, probably doesn't matter.

**Comment by oli-obk** [comment]
I'm assuming a `Vec` was used out of perf concerns. Could use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/sorted_map/struct.SortedMap.html instead to get similar behaviour.

But I don't think this is gonna cause perf issues either way (tho maybe check whether the original PR that added the code had any discussion on this)

**Comment by WaffleLapkin** [comment]
@oli-obk I'm not sure how `Vec`s would be any better.

- Previous impl for places used a hash set anyway, so it payed for collecting into a vec, and then *copied everything to a newly created hash set*.
- Previous impl for values used a sort+dedup which also doesn't look any better than a hash set.

**Comment by oli-obk** [comment]
> * Previous impl for values used a sort+dedup which also doesn't look any better than a hash set.

it does at small sizes 😆 the hashset only pays off after a certain number of entries.

**Pull request merged**